rpcs3/Emu/NP/upnp_handler.cpp

```diff
@@ -0,0 +1,182 @@
+#include "upnp_handler.h"
+#include "util/logs.hpp"
+
+#include <miniwget.h>
+#include <upnpcommands.h>
+
+LOG_CHANNEL(upnp_log, "UPNP");
+
+upnp_handler::~upnp_handler()
+{
+	std::lock_guard lock(m_mutex);
+
+	for (const auto& [protocol, prot_bindings] : m_bindings)
+	{
+		for (const auto& [internal_port, external_port] : prot_bindings)
+		{
+			remove_port_redir_external(external_port, protocol);
+		}
+	}
+}
+
+void upnp_handler::upnp_enable()
+{
+	std::lock_guard lock(m_mutex);
+
+	m_cfg.load();
+
+	auto check_igd = [&](const char* url) -> bool
+	{
+		int desc_xml_size = 0;
+		int status_code = 0;
+
+		m_igd_data = {};
+		m_igd_urls = {};
+
+		char* desc_xml = static_cast<char*>(miniwget(url, &desc_xml_size, 1, &status_code));
+
+		if (!desc_xml)
+			return false;
+
+		parserootdesc(desc_xml, desc_xml_size, &m_igd_data);
+		free(desc_xml);
+		desc_xml = nullptr;
+		GetUPNPUrls(&m_igd_urls, &m_igd_data, url, 1);
+
+		return true;
+	};
+
+	std::string dev_url = m_cfg.get_device_url();
+
+	if (!dev_url.empty())
+	{
+		if (check_igd(dev_url.c_str()))
+		{
+			upnp_log.notice("Saved UPNP(%s) enabled", dev_url);
+			m_active = true;
+			return;
+		}
+
+		upnp_log.error("Saved UPNP(%s) isn't available anymore", dev_url);
+	}
+
+	upnp_log.notice("Starting UPNP search");
+
+	int upnperror = 0;
+	UPNPDev* devlist = upnpDiscover(2000, nullptr, nullptr, 0, 0, 2, &upnperror);
+
+	if (devlist)
+	{
+		const UPNPDev* dev = devlist;
+		while (dev)
+		{
+			if (strstr(dev->st, "InternetGatewayDevice"))
+				break;
+			dev = dev->pNext;
+		}
+
+		if (dev)
+		{
+			int desc_xml_size = 0;
+			int status_code = 0;
+			char* desc_xml = static_cast<char*>(miniwget(dev->descURL, &desc_xml_size, 1, &status_code));
+
+			if (desc_xml)
+			{
+				IGDdatas igd_data{};
+				UPNPUrls igd_urls{};
+				parserootdesc(desc_xml, desc_xml_size, &igd_data);
+				free(desc_xml);
+				desc_xml = nullptr;
+				GetUPNPUrls(&igd_urls, &igd_data, dev->descURL, 1);
+
+				upnp_log.notice("Found UPnP device type:%s at %s", dev->st, dev->descURL);
+
+				m_cfg.set_device_url(dev->descURL);
+				m_cfg.save();
+
+				m_active = true;
+			}
+			else
+			{
+				upnp_log.error("Failed to retrieve UPNP xml for %s", dev->descURL);
+			}
+		}
+		else
+		{
+			upnp_log.error("No UPNP IGD device was found");
+		}
+
+		freeUPNPDevlist(devlist);
+	}
+	else
+	{
+		upnp_log.error("No UPNP device was found");
+	}
+}
+
+void upnp_handler::add_port_redir(std::string_view addr, u16 internal_port, std::string_view protocol)
+{
+	std::lock_guard lock(m_mutex);
+
+	if (!m_active)
+		return;
+
+	bool succeeded = false;
+	u16 external_port = internal_port;
+	std::string internal_port_str = fmt::format("%d", internal_port);
+
+	while (!succeeded)
```

Comment:
succeeded is never set to true... might as well just write while (true)
Also, this will do an endless loop if it never succeeds in the add command.
Maybe add a timeout or some max attempts